### PR TITLE
feat(CDC) Snapshot consumer bootstrap state

### PR DESCRIPTION
**Pull request by fpacifici**
This provides a working implementation for the bootstrap state for the snapshot consumer.

This is the state the snapshot consumer starts into. When we starts a consumer of a dataset that is able to load snapshot these things happen in sequence:
1) we consume the cdc_control topic to know if there is a pending snapshot we need to catch up with 
2) if there is no snapshot we commit the cdc_control and start consuming the main cdc topic with the batching consumer
3) if there is a pending snapshot, we would start in paused state and keep consuming the control topic waiting for the snapshot ready message
4) if we find a message that indicates the snapshot has been loaded, we switch to the cdc consumer but ignore all the transactions up to the snapshot itself.

This PR provides the logic that consumes the control topic, addresses corner cases (like overlapping snapshots), and decides which state to starts into.

Builds on top of #419  and #437

### CODE REVIEW

File: snuba/stateful_consumer/control_protocol.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    ],
+    "definitions": {
+        "base": {
+            "type": "object",
+            "properties": {
+                "snapshot-id": {
+                    "type": "string"
+                }
+            },
+            "required": ["event", "snapshot-id"],
```
Comment (tkaemming) [inline snuba/stateful_consumer/control_protocol.py]: Why is "event" part of this `required` list? It's not part of the property mapping, and seems like it's redundant with the definitions below?

File: snuba/stateful_consumer/control_protocol.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+            },
+            "required": ["temp_table"],
+        },
+        "snapshot-loaded": {
+            "allOf": [
+                {"$ref": "#/definitions/base"},
+                {
+                    "properties": {
+                        "event": {"const": "snapshot-loaded"},
+                        "datasets": {
```
Comment (tkaemming) [inline snuba/stateful_consumer/control_protocol.py]: Why is this plural? It looks like only one can be included. Why isn't this itself just a reference to the definition directly, rather than through an object type and `additionalProperties` below?

File: snuba/stateful_consumer/control_protocol.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                            "properties": {
+                                "xmin": {"type": "number"},
+                                "xmax": {"type": "number"},
+                                "xip-list": {
+                                    "type": "array",
+                                    "items": [
+                                        {"type": "number"}
+                                    ],
+                                },
+                            }
```
Comment (tkaemming) [inline snuba/stateful_consumer/control_protocol.py]: Shouldn't all of `xmin`/`xmax`/`xip-list` be required?

File: snuba/stateful_consumer/states/bootstrap.py [hunk 4]
```diff
@@ -1,25 +1,101 @@
...
+
+    def get_active_snapshot_msg(self) -> Optional[ControlMessage]:
+        return self.__active_snapshot_msg
+
+    def process_init(self, msg: SnapshotInit) -> None:
+        logger.debug("Processing init message for %r", msg.id)
+        if msg.product != settings.SNAPSHOT_LOAD_PRODUCT:
+            return
+        if self.__active_snapshot_msg:
+            if isinstance(self.__active_snapshot_msg, SnapshotInit):
```
Comment (tkaemming) [inline snuba/stateful_consumer/states/bootstrap.py]: When would this not be true? What type is `__active_snapshot_msg`?

Later: OK. I saw this assigned again below. Ideally the design of this class would be explained in the docstring or a comment or something, rather than requiring the reader to read the class first and figure it out for themself from the implementation

File: snuba/stateful_consumer/states/bootstrap.py [hunk 5]
```diff
@@ -1,25 +1,101 @@
...
+                    self.__active_snapshot_msg.id,
+                )
+                return
+
+        if msg.id in self.__processed_snapshots:
+            logger.warning(
+                "Duplicate Snapshot init: %r",
+                msg.id,
+            )
+        self.__processed_snapshots.add(msg.id)
```
Comment (tkaemming) [inline snuba/stateful_consumer/states/bootstrap.py]: Is "processed" really an accurate name for this? "Processed" seems to imply the entire snapshot process has been completed, and that doesn't seem to be the case here.

File: snuba/stateful_consumer/states/bootstrap.py [hunk 6]
```diff
@@ -38,10 +114,32 @@ def __init__(self,
...
+        self.__recovery_state = RecoveryState()
+
     def __handle_msg(self, message: Message) -> CommitDecision:
-        # TODO: Actually do something with the messages and drive the
-        # state machine to the next state.
-        return CommitDecision.DO_NOT_COMMIT
+        value = json.loads(message.value())
+        parsed_message = parse_control_message(value)
+
+        current_snapshot = self.__recovery_state.get_active_snapshot_msg()
```
Comment (tkaemming) [inline snuba/stateful_consumer/states/bootstrap.py]: This assignment doesn't have anything to do with the code immediately following it, it should be moved down closer to the usage unless there's some reason it needs to be here (if there is, it should be commented.)

File: snuba/stateful_consumer/states/bootstrap.py [hunk 7]
```diff
@@ -38,10 +114,32 @@ def __init__(self,
...
+            self.__recovery_state.process_snapshot_loaded(
+                parsed_message,
+            )
+
+        new_snap = self.__recovery_state.get_active_snapshot_msg()
+        if new_snap is None:
+            logger.debug("Committing offset %r ", message.offset())
+            return CommitDecision.COMMIT_THIS
+        elif current_snapshot is None or new_snap.id != current_snapshot.id:
+            logger.debug("Committing previous offset to %r ", message.offset())
```
Comment (tkaemming) [inline snuba/stateful_consumer/states/bootstrap.py]:
```suggestion
            logger.debug("Committing previous offset to %r", message.offset())
```

File: snuba/stateful_consumer/states/bootstrap.py [hunk 8]
```diff
@@ -38,10 +114,32 @@ def __init__(self,
...
+        elif isinstance(parsed_message, SnapshotAbort):
+            self.__recovery_state.process_abort(parsed_message)
+        elif isinstance(parsed_message, SnapshotLoaded):
+            self.__recovery_state.process_snapshot_loaded(
+                parsed_message,
+            )
+
+        new_snap = self.__recovery_state.get_active_snapshot_msg()
+        if new_snap is None:
+            logger.debug("Committing offset %r ", message.offset())
```
Comment (tkaemming) [inline snuba/stateful_consumer/states/bootstrap.py]:
```suggestion
            logger.debug("Committing offset %r", message.offset())
```

**Review by tkaemming — CHANGES_REQUESTED** [review]
Kind of struggling to see how the pieces fit together while reading this. I think elaborating on the design in the docstring or comments would make reviewing this now and understanding it in the future a lot more approachable.

File: snuba/consumers/strict_consumer.py [hunk 9]
```diff
@@ -142,7 +142,7 @@ def run(self) -> None:
                         topic=message.topic(),
                         offset=prev_watermark,
                     )
-                    self.__consumer.commit(commit_pos, asynchronous=False)
+                    self.__consumer.commit(offsets=[commit_pos], asynchronous=False)
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/consumers/strict_consumer.py]: It was supposed to be part of the previous PR, but none was using this code there.

File: snuba/stateful_consumer/control_protocol.py (hunk 1 above)
Comment (fpacifici) [reply snuba/stateful_consumer/control_protocol.py]: I think I just forgot it there. It took me a while to get this schema right, Likely I missed this one.

File: snuba/stateful_consumer/control_protocol.py (hunk 2 above)
Comment (fpacifici) [reply snuba/stateful_consumer/control_protocol.py]: When I designed the protocol (before doing any test around groupedmessage dataset) I was expecting a consumer to be able to reload multiple datasets at a time. 
This is not likely to be the case, but in the PR above, I mentioned I want to entirely remove this field.

File: snuba/stateful_consumer/control_protocol.py (hunk 3 above)
Comment (fpacifici) [reply snuba/stateful_consumer/control_protocol.py]: just forgot

File: snuba/stateful_consumer/states/bootstrap.py (hunk 4 above)
Comment (fpacifici) [reply snuba/stateful_consumer/states/bootstrap.py]: Ok, I will beef up the docstring

File: snuba/stateful_consumer/states/bootstrap.py [hunk 10]
```diff
@@ -1,25 +1,101 @@
...
+        return self.__active_snapshot_msg
+
+    def process_init(self, msg: SnapshotInit) -> None:
+        logger.debug("Processing init message for %r", msg.id)
+        if msg.product != settings.SNAPSHOT_LOAD_PRODUCT:
+            return
+        if self.__active_snapshot_msg:
+            if isinstance(self.__active_snapshot_msg, SnapshotInit):
+                logger.error(
+                    "Overlapping snapshots. Ignoring. Running %r. Init received %r.",
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/stateful_consumer/states/bootstrap.py]: will clarify

**Review by fpacifici — COMMENTED** [review]
I will add more details about the design in the RecoveryState docstring. Is that what you are referring to specifically . or is there something else that is not clear ?

File: snuba/stateful_consumer/states/bootstrap.py (hunk 6 above)
Comment (fpacifici) [reply snuba/stateful_consumer/states/bootstrap.py]: The lines below willchange the result of get_active_snapshot_msg, so this has to be before. It will be compared with new_snap later

**Comment by fpacifici** [comment]
Updates:
- added more documentation to the bootstrap state
- changed the way RecoveryState keeps track of the state itself. This should be more understandable
- Removed datasets from the snapshot-loaded message and added tables to snapshot-init so that the bootstrap state can ignore the tables it is not interested into.

File: snuba/stateful_consumer/states/bootstrap.py [hunk 11]
```diff
@@ -1,25 +1,178 @@
...
+        if msg.id in self.__past_snapshots:
+            logger.warning(
+                "Duplicate Snapshot init: %r",
+                msg.id,
+            )
+            return self.__commit_if_no_active_snapshot()
+        if self.__active_snapshot_id and not self.__loaded_snapshot_transactions:
+            logger.error(
+                "A new snapshot-init message was found before a previous one "
+                " was completed. Ignoring. Running %r. Init received %r.",
```
Comment (tkaemming) [inline snuba/stateful_consumer/states/bootstrap.py]: "running" here still strikes me as odd, since it doesn't change the active snapshot. Would it make more sense to word this as `Ignoring %r, and continuing with %r` instead?

It also seems a bit weird to use `error` when this isn't logging an exception, kind of feels like this should be a `warning` instead.

File: snuba/stateful_consumer/states/bootstrap.py [hunk 12]
```diff
@@ -1,25 +1,178 @@
...
+        """
+        Processes an abort message found on the control topic. If this comes after
+        a snapshot-init message, it makes the state machine drop that snapshot.
+        """
+        logger.debug("Processing abort message for %r", msg.id)
+        if msg.id not in self.__past_snapshots:
+            return self.__commit_if_no_active_snapshot()
+        if self.__active_snapshot_id != msg.id:
+            logger.warning(
+                "Aborting a snapshot that is not active. Active %r, Abort %r",
```
Comment (tkaemming) [inline snuba/stateful_consumer/states/bootstrap.py]: I think here (and in `process_snapshot_below` it might make sense to reword this as noted above (if you agree with my interpretation of what this is saying.)

**Review by tkaemming — APPROVED** [review]
This is a lot easier to understand now, thanks!

There's some minor things inline (mostly typos or wording changes and a couple style things) but nothing serious or blocking.

File: snuba/stateful_consumer/control_protocol.py [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+                    "type": "string"
+                }
+            },
+            "required": ["snapshot-id"],
+        },
+        "snapshot-init": {
+            "allOf": [
+                {"$ref": "#/definitions/base"},
+                {
+                    "properties": {
```
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/stateful_consumer/control_protocol.py]: Same as here: https://github.com/getsentry/snuba/pull/441/files/090d0499a517fc2cffff244c9a3be69f51f22dfe#diff-b4875ea4684246d1a462c8d94f40ec6dR24
